### PR TITLE
Issue #111 - Cleanup Utility Module

### PR DESCRIPTION
**Pull request by andrewnguyen22**
# Objective
Reduce the code footprint of the utility module, so it is more understandable to new readers while finding a balance between minimizing legacy complexity in the future and not over-generalizing the existing code.

# Origin Document

PR73 ([First iteration of a PostgreSQL based Persistence Schema](https://github.com/pokt-network/pocket/pull/73)) underwent several iterations in order to make the [persistence module implementation](https://github.com/pokt-network/pocket/tree/main/persistence) have the optionality to remain verbose on a per-actor basis but also avoid code duplication and redundancy.

The first iteration of the [utility module](https://github.com/pokt-network/pocket/tree/main/persistence) did not undergo the same iterations and therefore has some low-hanging fruit to simplify the codebase before further functionality is added.

# Goals / Deliverables
- [x] Extract common actor functionality into general interfaces
- [x] Reduce the code complexity and code footprint so it is more approachable and maintainable
- [ ] Identify and document/cleanup parts of the code that are unclear to new readers
- [x] Generalize both unit tests and functionality where possible
- [x] Maintain verbose entry points for each actor
General:
- [x] Update the CHANGELOG
- [x] Update the README
- [x] Update the global documentation & references
- [x] Document small issues / TODOs along the way

## Testing Methodology
 - **Utility tests**: `make test_utility_*`
 - **All tests**: `make test_all`
 - **LocalNet**: works following the instructions in [docs/development](https://github.com/pokt-network/pocket/tree/main/docs/development)

## Non-goals
- Adding any new functionality

---

## Changes Made
- Removed transaction fees from the transaction structure as fees will be enforced at the state level
- Removed actor specific messages (besides DoubleSign) and added actorType field to the struct
- Removed pause messages and functionality as it is out of scope for the current POS iteration
- Removed session and test-scoring as it's out of scope for the current POS iteration
- Consolidated unit test functionality for actors
- Modified pre-persistence to match persistence for Update(actor), 'amountToAdd' is now just 'amount'

## Future work
- There's more documentation that is needed for this to be complete. In a separate issue, I'd like to tackle connecting Utility with Persistence.
- Further cleanup / refactoring / optimizations can be made, but that can be left to future iteration of the module.

---

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>
Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MacBook-Pro-2.local>
Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MBP-2.lan>
Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

### CODE REVIEW

**Comment by Olshansk** [comment]
Discussed offline in a DM but also posting here for transparency for others.

@andrewnguyen22 Looking at the `Makefile`, it seems like we need to either merge/rebase (whichever is easier in this scenario) on top of `main` to view the changes.

**Comment by andrewnguyen22** [comment]
Merged into main, should be good to go

File: persistence/pre_persistence/app_test.go
Comment (Olshansk) [inline persistence/pre_persistence/app_test.go:118]: Add back or remove?

Just leaving one comment but there are a few commented out test sections, and not sure whether we need to keep them in.

File: shared/tests/utility_module/actor_test.go
Comment (Olshansk) [inline shared/tests/utility_module/actor_test.go:24]: All the tests in `actor_test.go` are still named `TestUtilityContext_*App`.

Is it intentional? If not, please update

File: shared/tests/utility_module/actor_test.go
Comment (Olshansk) [inline shared/tests/utility_module/actor_test.go:86]: DItto fro all other Get calls in the file.

Should this be `GetAllTestingApps`  or something more general?

File: utility/proto/message.proto
Comment (Olshansk) [inline utility/proto/message.proto:9]: Can we also make `StakingStatus` an enum? It's something I get confused about quite often throughout...

File: utility/proto/message.proto
Comment (Olshansk) [inline utility/proto/message.proto:27]: Optional was removed in proto3, but seems like it's an experimental feature again according to https://github.com/protocolbuffers/protobuf/blob/main/docs/implementing_proto3_presence.md.

If we're adding this back, can you pass `--experimental_allow_proto3_optional` in the Makefile? I think it's a noop otherwise

Screen Shot 2022-07-25 at 12 16 26 PM

File: utility/actor.go
Comment (Olshansk) [inline utility/actor.go:21]: NIT: Can you make `actorType typesUtil.ActorType` the first param so it's easier to read during usage?

Ditto below

**Review by Olshansk — CHANGES_REQUESTED** [review]
Screen Shot 2022-07-25 at 12 19 07 PM

Love that delta :)

---

@andrewnguyen22 We can potentially do a handoff on this, but I think there's still some "low hanging fruit" to refactor before I take it over. Primarily asking you to do this so I can focus on persistence for now.

1. Left some minor comments/questions inline.

2. Under `shared/tests/utility_module/*_test.go` we make use of `t.Fatal` quite a bit, which I believe can be eliminated completely by leveraging the `require` package. It'll be easier to read and also gives better error messages since it output expected and actual by default with an optional message.

For example, in `validator_test.go`, the following:

```
	got, _, err := ctx.(*PrePersistenceContext).GetValidator(actor.Address, height)
	require.NoError(t, err)
	if !bytes.Equal(actor.Address, got.Address) || !bytes.Equal(actor.PublicKey, got.PublicKey) {
		t.Fatalf("unexpected actor returned; expected %v got %v", actor, got)
	}
```
can be changed to:

```
	require.Equal(t, actor.Address, got.Address, "Unexpected validator address")
	require.Equal(t, actor.PublicKey, got.PublicKey, "Unexpected validator public key")
```

File: shared/tests/utility_module/actor_test.go
Comment (andrewnguyen22) [reply shared/tests/utility_module/actor_test.go:86]: I'm using apps here, but could be any actor

File: utility/proto/message.proto
Comment (andrewnguyen22) [reply utility/proto/message.proto:9]: Not worth it, it will be deprecated soon

**Comment by Olshansk** [comment]
@andrewnguyen22 Per an offline discussion, it is likely you have mostly completed this work, but I wanted to share what my personal approach is to replacing lots of code using regexes.

_Disclaimer: There's probably always a better way to do it so this is just knowledge sharing._

I use **case sensitive regex find & replace all** in vscode* and create a few pattern (see two examples below) where I use the matches from the find (i.e. `(.*)`) in the replace (i.e. `$1`, `$2`, `$3`). It's a fun way to practice regex and helps alleviate my allergy of doing repetitive copy-pasta work though I realize it may be fun/meditative to some.

_* I'd be surprised if Goland or other IDEs don't have this feature_

**Example1**
Screen Shot 2022-07-26 at 2 48 47 PM

**Example2**
Screen Shot 2022-07-26 at 2 43 54 PM

**Review by Olshansk — CHANGES_REQUESTED** [review]
@andrewnguyen22 Tagging this off back to you since I'm getting really tired today...

See the changes I made by running `git diff 38684afcc47089802d023286de575a1c4787143f` after pulling.

0. **NOTE**: I apologize if some of the comments I left on the PR inline may not have been very descriptive. Feel free to ignore or ask for a clarification if it doesn't make sense.

1. **Minor Actionable**: There are still a few `t.fatal` if you search for it under the `utility` directory. PTAL and update

2. **FYI**: I think there is still opportunity to simplify `actor.go` but I don't have an idea for it yet. Can you add a TODO on me in the main README for utility?

3. **Request**: Can you add a "code structure" explanation in the README? Similar to the one that's in the currently outstanding P2P PR?

4. **Major Actionable**: The only big action item I think that's left is updating `actor_test.go` to use generics and potentially testing for all actors. For example, we currently use `GetAllTestingApps` to test every functionality for apps, even though we use `actor` inline. I think there's a very easy opportunity to create a map from ActorType->GetAllTestingActors (App -> GetApps, Val -> GetVal, etc...) and execute every single test for every actor to make sure none of them break.

I haven't coded this out yet myself, but believe it can be possible using generics, and golang's testing structure

5. **Question**: Can we remove `vote.proto` altogether? It's not really being used in the Consensus module? Ditto for `DoubleSign`?

File: utility/module.go
Comment (andrewnguyen22) [reply utility/module.go:47]: We do not

File: utility/module.go
Comment (andrewnguyen22) [reply utility/module.go:17]: I'm not sure the actionable here. gofmt formats it this way

File: utility/module.go
Comment (andrewnguyen22) [reply utility/module.go:55]: Leave that to the persistence #128

File: utility/module.go
Comment (andrewnguyen22) [reply utility/module.go:99]: No

File: utility/proto/vote.proto
Comment (andrewnguyen22) [reply utility/proto/vote.proto:7]: Yes

Comment (andrewnguyen22) [reply utility/proto/vote.proto:7]: No, it's used in doubleSign Messge

**Comment by andrewnguyen22** [comment]
Didn't mean to request a review yet

**Comment by andrewnguyen22** [comment]
> Didn't mean to request a re-review yet

Good to go

File: utility/module.go
Comment (Olshansk) [reply utility/module.go:17]: tl;dr Ignore.

Line 22 had `// TODO: Add `maxTransactionBytes` and `maxTransactions` to cfg.Utility`, figured we could do it here but it's out of scope.

**Review by phthan0 — APPROVED** [review]
Looks good.

**Comment by Olshansk** [comment]
I'm still in the process of making some changes that I plan to push up by EOD. Please hold off on merging for now

**Comment by Olshansk** [comment]
@andrewnguyen22 We'll definitely have another iteration on this at some point, but for now I think this is a major step forward and improvement.

I made additional changes, merged with main, ran the unit tests, tested local cloud and updated the description.

Feel free to squash & merge if it looks good to you.

**Review by Olshansk — APPROVED** [review]

**Pull request merged**
